apps/music/js/metadata/ogg.js

```diff
@@ -39,131 +43,194 @@ var OggMetadata = (function() {
    * @return {Promise} A Promise returning the parsed metadata object.
    */
   function parse(blobview) {
-    readIdentificationHeader(blobview);
-    return readCommentHeader(blobview);
+    var parser = new VorbisCommentParser();
+    parser.readIdentificationHeader(blobview);
+    return parser.readVorbisComment(blobview);
   }
 
   /**
-   * Read the identification header of an Ogg container. It's always the first
-   * page.
+   * Constructor for the VorbisComment parser.
    *
-   * @param {BlobView} page The audio file being parsed.
+   * @param {BlobView} blobview The audio file to parse.
+   * @param {Metadata} metadata The (partially filled-in) metadata object.
    */
-  function readIdentificationHeader(page) {
-    var header = readPageHeader(page);
-    if (header.segment_table.length !== 1) {
-      throw new Error(
-        'ogg identification header expected as only packet of first page'
-      );
-    }
-    // Skip over the identification header.
-    page.advance(header.segment_table[0]);
+  function VorbisCommentParser() {
+    this.continued = false;
+    this.read_comments = 0;
+    return this;
   }
 
-  /**
-   * Read the comment header of an Ogg container.
-   *
-   * @param {BlobView} page The audio file being parsed.
-   * @return {Promise} A Promise that resolves with the completed metadata.
-   */
-  function readCommentHeader(page) {
-    var metadata = {};
-    var header = readPageHeader(page);
+  VorbisCommentParser.prototype = {
+    /**
+     * Read the identification header of an Ogg container. It's always the first
+     * page.
+     *
+     * @param {BlobView} blobview The audio file being parsed.
+     */
+    readIdentificationHeader: function(blobview) {
+      var header = this.readPageHeader(blobview);
+      if (header.segment_table.length !== 1) {
+        throw new Error(
+          'ogg identification header expected as only packet of first page'
+        );
+      }
+      // Skip over the identification header.
+      blobview.advance(header.segment_table[0]);
+    },
+
+    /**
+     * Read the Vorbis Comment. Deal with multiple pages. Parse the
+     * picture block too.
+     *
+     * @param {BlobView} page The audio file being parsed.
+     * @return {Promise} A Promise that resolves with the completed metadata.
+     */
+    readVorbisComment: function(blobview) {
+      return new Promise(function(resolve, reject) {
+
+        this.done = function(metadata) {
+          LazyLoader.load('js/metadata/vorbis_picture.js').then(() => {
+            return VorbisPictureComment.parsePictureComment(metadata);
+          }).then(resolve);
+        }.bind(this);
+
+        this.readCommentHeader(blobview);
```

Comment:
Since this function should be returning a Promise, you should return the result of it so that the caller can wait for resolution of this inner Promise.